### PR TITLE
No BOM for serialized output

### LINKED ISSUE #53: UTF-8 BOM (Byte Order Mark) is incorrectly added to the serialized message bodies
## Who's affected

Anyone manually deserializing NServiceBus messages using custom code.

## Symptoms

1. The first 3 bytes of the serialized message body are BOM bytes (`0xEF`,`0xBB`,`0xBF`)
1. An attempt to deserialize messages using `JsonConvert.DeserializeObject()` throws an exception "Newtonsoft.Json: Unexpected character encountered while parsing value: ?. Path '', line 0, position 0."

## Backported to

- 1.2.1 https://github.com/Particular/NServiceBus.Newtonsoft.Json/pull/56

## Original issue

The BOM for UTF-8 is currently added to outgoing messages:

```c#
_messageSession.Send(new SaySomethingCommand {Message = "Hello World"});
```

Becomes:

```json
{"Message":"Hello World"}
```

However, looking at the bytes, the BOM is added:

```
EFBBBF7B224D657373616765223A2248656C6C6F20576F726C64227D
```

Which, decoding, looks like:

```
ï»¿{"Message":"Hello World"}
```

If you try to manually deserialize through JSON.NET, say for an integration scenario, Azure Functions, Web Jobs etc:

```c#
var body = Encoding.UTF8.GetString(message.Body);
var command = JsonConvert.DeserializeObject<SaySomethingCommand>(body);
```

This will fail with the message:

```
Newtonsoft.Json: Unexpected character encountered while parsing value: ?. Path '', line 0, position 0.
```

Per the [RFC](https://tools.ietf.org/html/rfc7159#section-8.1):

> Implementations MUST NOT add a byte order mark to the beginning of a JSON text.

The JSON in the message, because it includes a BOM, isn't actually valid JSON.

### PR DESCRIPTION
**Pull request by jbogard**
Changing default setting of encoding to not include the BOM for UTF8 per JSON RFC spec. Fixes #53

### CODE REVIEW

File: src/NServiceBus.Newtonsoft.Json.sln.DotSettings [hunk 1]
```diff
@@ -557,6 +557,9 @@ II.2.12 &lt;HandlesEvent /&gt;&#xD;
 	<s:String x:Key="/Default/CodeStyle/Naming/XamlNaming/UserRules/=XAML_005FFIELD/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
 	<s:String x:Key="/Default/CodeStyle/Naming/XamlNaming/UserRules/=XAML_005FRESOURCE/@EntryIndexedValue">&lt;Policy Inspect="True" Prefix="" Suffix="" Style="AaBb" /&gt;</s:String>
 	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpFileLayoutPatternsUpgrade/@EntryIndexedValue">True</s:Boolean>
+	<s:Boolean x:Key="/Default/Environment/SettingsMigration/IsMigratorApplied/=JetBrains_002EReSharper_002EPsi_002ECSharp_002ECodeStyle_002ECSharpKeepExistingMigration/@EntryIndexedValue">True</s:Boolean>
```
Comment (SeanFeldman) [inline src/NServiceBus.Newtonsoft.Json.sln.DotSettings]: Could you please undo this change?

Comment (jbogard) [reply src/NServiceBus.Newtonsoft.Json.sln.DotSettings]: Yep.

File: src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs [hunk 2]
```diff
@@ -16,6 +16,7 @@ class JsonMessageSerializer : IMessageSerializer
         Func<Stream, JsonReader> readerCreator;
         Func<Stream, JsonWriter> writerCreator;
         NewtonSerializer jsonSerializer;
+        static readonly Encoding utf8NoBOM = new UTF8Encoding(false);
```
Comment (SeanFeldman) [inline src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: According to [documentation](https://docs.microsoft.com/en-us/dotnet/api/system.text.utf8encoding.-ctor?view=netframework-4.7.2#System_Text_UTF8Encoding__ctor_System_Boolean_):
> For security reasons, you should enable error detection by calling a constructor that includes a `throwOnInvalidBytes` parameter and setting its value to `true`.

Comment (bording) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: That would suggest we'd want the overload that takes two bools, but if you look at the `Encoding.UTF8` [docs](https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.utf8?view=netframework-4.7.2#remarks), it turns invalid bytes into `?` instead.

It seems that the versions that you can create don't have that option. They either ignore invalid bytes or throw.

Neither option seems to be a good fit here, so I'm not sure what we can do here.

Comment (jbogard) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: If you didn't supply an encoder, this is what it would use:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/io/streamwriter.cs#L112

Which then defaults to false:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/text/utf8encoding.cs#L77

I guess to preserve the existing behavior, it doesn't throw and puts in ?'s. In the interest of preserving behavior, I'd see that.

I can add a test for this scenario, though.

Comment (bording) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: @jbogard It's not clear to me what you're suggesting. At least from the docs, it appears that there are three different behaviors regarding how invalid bytes are handled:
- `Encoding.UTF8`: replace with `?`
- `new UTF8Encoding (false, false)` : "ignore them"
- `new UTF8Encoding (false, true)`: "throw exception"

If that is the case, then this seems like a problem since it would cause an observable difference in how we handle them after this PR, and that gets into wire compat concerns.

Perhaps the `(false, false)` case is actually the same as `Encoding.UTF8` despite the docs not making that clear?

Comment (jbogard) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: If you *don't* supply an encoder, it's `(false, true)`. If you use `Encoding.UTF8`, it's `(true, false)`:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/text/encoding.cs#L1549

To preserve the existing behavior on throwing, we'd want `false` for the second parameter, for writing. I left the existing reading encoder as-is.

So I think we'd want `(false, false)`, changing the BOM for writing but leaving the 2nd argument as-is.

Comment (jbogard) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: I'm going to add the second argument to make it explicit.

Comment (jbogard) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: For posterity, `StreamReader` defaults to `Encoding.UTF8`:

https://github.com/Microsoft/referencesource/blob/master/mscorlib/system/io/streamreader.cs#L138

Which is `(true, false)`.

Comment (jbogard) [reply src/NServiceBus.Newtonsoft.Json/JsonMessageSerializer.cs]: Am I supposed to mark the conversation as resolved?

**Comment by HEskandari** [comment]
@jbogard Thanks for the PR. We're still working on testing backwards and wire compatibility tests. We'll merge it once that's done.

File: src/Tests/No_UTF8_BOM.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
+﻿using System.IO;
+using System.Text;
+using NServiceBus.MessageInterfaces.MessageMapper.Reflection;
+using NServiceBus.Newtonsoft.Json;
+using NUnit.Framework;
+
+[TestFixture]
+public class No_UTF8_BOM
```
Comment (bording) [inline src/Tests/No_UTF8_BOM.cs]: Let's rename this to `When_serializing_a_message`

File: src/Tests/No_UTF8_BOM.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+{
+
+    [Test]
+    public void Run()
```
Comment (bording) [inline src/Tests/No_UTF8_BOM.cs]: Let's rename this to `Should_not_emit_UTF8_BOM`

File: src/Tests/With_UTF8_BOM.cs [hunk 5]
```diff
@@ -0,0 +1,33 @@
+﻿using System.IO;
+using System.Text;
+using NServiceBus.MessageInterfaces.MessageMapper.Reflection;
+using NServiceBus.Newtonsoft.Json;
+using NUnit.Framework;
+
+[TestFixture]
+public class With_UTF8_BOM
```
Comment (bording) [inline src/Tests/With_UTF8_BOM.cs]: Let's rename this to `When_deserializing_a_message`

File: src/Tests/With_UTF8_BOM.cs [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+{
+
+    [Test]
+    public void Run()
```
Comment (bording) [inline src/Tests/With_UTF8_BOM.cs]: Let's rename this to `Should_handle_message_with_UTF8_BOM`

Comment (bording) [reply src/Tests/With_UTF8_BOM.cs]: I'd also like to see a Should_handle_message_without_UTF8_BOM test.

File: src/Tests/With_UTF8_BOM.cs [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        var messageMapper = new MessageMapper();
+        var serializer = new JsonMessageSerializer(messageMapper, null, null, null, null);
+
+        var serialized = new UTF8Encoding(true).GetBytes($"{{\"{nameof(SimpleMessage.SomeProperty)}\":\"John\"}}");
```
Comment (bording) [inline src/Tests/With_UTF8_BOM.cs]: When I debug the test and look at the contents of this array, it doesn't actually have a BOM, so this test currently isn't testing what it's saying it does.

File: src/Tests/With_UTF8_BOM.cs (hunk 6 above)
Comment (jbogard) [reply src/Tests/With_UTF8_BOM.cs]: Added.

File: src/Tests/With_UTF8_BOM.cs (hunk 5 above)
Comment (jbogard) [reply src/Tests/With_UTF8_BOM.cs]: Done.

File: src/Tests/No_UTF8_BOM.cs (hunk 4 above)
Comment (jbogard) [reply src/Tests/No_UTF8_BOM.cs]: Done.

File: src/Tests/No_UTF8_BOM.cs (hunk 3 above)
Comment (jbogard) [reply src/Tests/No_UTF8_BOM.cs]: Done.

**Comment by SeanFeldman** [comment]
This PR is targeting the `develop` branch. New PRs were raised against `master` to hotfix the latest (https://github.com/Particular/NServiceBus.Newtonsoft.Json/pull/55) and backported to the supported previous version (https://github.com/Particular/NServiceBus.Newtonsoft.Json/pull/56). Commits from this PR were used. 

Thank you @jbogard for your contribution.
